File: puzzle15.java
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Scanner;

public class puzzle15 {
    static PriorityQueue<Node> queue;
    static final int N = 4;
   // static int counter = 0;
    static Node finalState;
    public static void main(String[] args) {
        int i, j;
        //int start[] = {5,1,2,4,9,6,3,8,13,10,7,11,0,14,15,12};
        //int hard[] = {11, 5, 12, 14, 15, 2, 0, 9, 13, 7, 6, 1, 3, 10, 4, 8};
        //int hard[] = {6, 1, 3, 4, 2, 7, 15, 8, 5, 10, 9, 11, 13, 14, 12, 0};
        int end[] = {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,0};
        Compare compare = new Compare();
        queue = new PriorityQueue<>(compare);
        Node initialState = new Node();
        Scanner scanner = new Scanner(System.in);
        finalState = new Node();
        System.out.println("Enter Initial State(0 for empty):");
        for(i = 0; i < N; i++){
            for(j = 0; j < N; j++){
                initialState.state[i][j] = scanner.nextInt();
                //initialState.state[i][j] = hard[i*4+j];
                if(initialState.state[i][j]==0){
                    initialState.xCord = i;
                    initialState.yCord = j;
                }
            }
        }
        for(i = 0; i < N; i++){
            for(j = 0; j < N; j++){
                finalState.state[i][j] = end[i*4+j];
            }
        }
        //PRINT.ArrayMatrix(initialState.state);
        //PRINT.ArrayMatrix(finalState.state);
        initialState.level = 0;
        initialState.setCost(finalState);

        queue.add(initialState);
        System.out.println("And the search begins: ");
        findSolution();
    }

    // private static void display(Node parent){
    //     Scanner scanner = new Scanner(System.in);
    //     counter++;
    //     System.out.println(counter+".\nLevel: "+ parent.level);
    //     PRINT.ArrayMatrix(parent.state);
    //     System.out.println("Cost: "+parent.cost);
    //     System.out.println("Path Taken: "+ parent.pastDirections.toString());
    //     //scanner.next();
    // }

    private static void findSolution() {
        int bonusPoints = 0;
        while (!queue.isEmpty()){
            Node parent = queue.remove();
            Node[] temp = new Node[4];
            char lastDirection;
            if(parent.pastDirections.length()>0)
                lastDirection = parent.pastDirections.charAt(parent.pastDirections.length()-1);
            else
                lastDirection = ' ';
            //display(parent);
            if(parent.xCord>0 && lastDirection!='D'){                 //UP
                temp[0] = new Node(parent, 'U');
                temp[0].setCost(finalState);
                if(temp[0].cost == temp[0].level-bonusPoints)
                    solved(temp[0]);
                queue.add(temp[0]);
            }
            if(parent.xCord<(N-1) && lastDirection!='U'){                 //DOWN
                temp[1] = new Node(parent, 'D');
                temp[1].setCost(finalState);
                if(temp[1].cost == temp[1].level-bonusPoints)
                    solved(temp[1]);
                queue.add(temp[1]);
            }
            if(parent.yCord>0 && lastDirection!='R'){                 //LEFT
                temp[2] = new Node(parent, 'L');
                temp[2].setCost(finalState);
                if(temp[2].cost == temp[2].level-bonusPoints)
                    solved(temp[2]);
                queue.add(temp[2]);
            }
            if(parent.yCord<(N-1) && lastDirection!='L'){                 //RIGHT
                temp[3] = new Node(parent, 'R');
                temp[3].setCost(finalState);
                if(temp[3].cost == temp[3].level-bonusPoints)
                    solved(temp[3]);
                queue.add(temp[3]);
            }
        }
        System.out.println("No possible combination found!");
    }

    private static void solved(Node temp) {
        System.out.println("Puzzle solved in "+temp.level+ " steps.\nMoves made: "+temp.pastDirections.toString());
        System.exit(0);
    }
}
class Node{
    int state[][], cost, level, xCord, yCord;
    StringBuffer pastDirections;
    Node(){
        state = new int[4][4];
        cost = level = 0;
        pastDirections = new StringBuffer(100);
    }
    Node(Node parent, char direction){
        state = new int[4][4];
        copyState(parent);
        level = parent.level+1;
        cost = 0;
        pastDirections = new StringBuffer(100);
        pastDirections.append(parent.pastDirections);
        pastDirections.append(direction);
        if(direction=='U'){
            xCord = parent.xCord-1;
            yCord = parent.yCord;
        }
        else if(direction=='D'){
            xCord = parent.xCord+1;
            yCord = parent.yCord;
        }
        else if(direction=='L'){
            xCord = parent.xCord;
            yCord = parent.yCord-1;
        }
        else{
            xCord = parent.xCord;
            yCord = parent.yCord+1;
        }
        state[parent.xCord][parent.yCord] = state[xCord][yCord];
        state[xCord][yCord] = 0;
    }
    void copyState(Node from){
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                state[i][j] = from.state[i][j];
            }
        }
    }
    void setCost(Node finalState){
        cost = level;
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                if(state[i][j] != finalState.state[i][j])
                    cost++;
            }
        }
        //Extra
/*
        if(state[0][0]==1){           //ROW 1
            cost-=100;
            if (state[0][1]==2){
                cost-=90;
                if(state[0][2]==3 && state[0][3]==4){         //Ist Row correct
                    cost-=100;
                    if(state[1][0]==5){
                        cost-=100;
                        if(state[1][1]==6){
                            cost-=90;
                            if(state[1][2]==7 && state[1][3]==8) {               //2nd Row Correct
                                cost-=100;
                                if(state[2][0]==9 && state[3][0]==13){          //1st Column Correct
                                    cost-=100;
                                    if(state[2][1]==10 && state[3][1]==14)          //2nd Column Correct
                                        cost-=100;
                                }
                            }
                        }
                    }
                }
            }
        }*/
    }
}

class Compare implements Comparator<Node>{
    @Override
    public int compare(Node o1, Node o2) {
        return o1.cost-o2.cost;
    }
}

File: puzzle15.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#define N 4
typedef struct Node{
   int cost, level, blankX, blankY;
   char *pastDirections;
   int state[N][N];
}NODE;

typedef struct heap{
  NODE arr[10000];
  int size;
}HEAP;

void swap(HEAP *heap, int i, int j){
  NODE temp;
  temp = heap->arr[i];
  heap->arr[i] = heap->arr[j];
  heap->arr[j] = temp;
}

void heapifyUp(HEAP *heap, int i){
  if(i==0)
    return;
  int parent = (i%2==0)?((i/2)-1):(i/2);
  if(heap->arr[i].cost>heap->arr[parent].cost)
    return;
  swap(heap, i , parent);
  heapifyUp(heap, parent);
}

void enqueue(HEAP *heap, NODE node){
  heap->arr[heap->size] = node;
  heapifyUp(heap, heap->size);
  heap->size += 1;
}

void heapifyDown(HEAP *heap, int i){
  int l = 2*i+1, r = 2*i + 2, max;
  if((l<heap->size) && (heap->arr[l].cost<heap->arr[i].cost))
    max = l;
  else
    max = i;

  if((r<heap->size) && (heap->arr[r].cost<heap->arr[max].cost))
    max = r;

  if(max != i){
    swap(heap, i, max);
    heapifyDown(heap, max);
  }
}

NODE dequeue(HEAP *heap){
  NODE t = heap->arr[0];
  heap->arr[0] = heap->arr[heap->size-1];
  heap->size-=1;
  heapifyDown(heap, 0);
}

void display(NODE node){
  int i,j;
  for(i=0;i<N;i++){
    for(j=0;j<N;j++)
      printf("%2d ",node.state[i][j]);
    printf("\n");
  }
}

void copyState(NODE *from, NODE *to){
  int i, j;
  for(i = 0; i < N; i++){
    for(j = 0; j < N; j++){
      to->state[i][j] = from->state[i][j];
    }
  }
}

void solved(NODE node){
  printf("The puzzle is solved in %d steps.\n", node.level);
  display(node);
  printf("The steps taken to solve this puzzle is: %s\n", node.pastDirections);
}

void setCost(NODE *node, NODE finalState){
  int i,j;
  printf("setCost %d %d\n", node->cost, node->level);
  node->cost = node->level;
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      if(node->state[i][j]!=finalState.state[i][j])
        node->cost++;
    }
  }
}

void setNode(NODE *temp, NODE *current, char *str){
  temp->level = current->level + 1;
  temp->pastDirections = "";
  strcat(temp->pastDirections, current->pastDirections);
  strcat(temp->pastDirections, str);
  copyState(current, temp);
  temp->state[current->blankX][current->blankY] = temp->state[temp->blankX][temp->blankY];
  temp->state[temp->blankX][temp->blankY] = 0;
}

void solveProblem(HEAP *heap, NODE finalState){
  printf("%d\n", heap->size);
  if(heap->size==0){
    printf("No possible number of moves can restore the inital state to the final state.\n");
    return;
  }
  NODE current = dequeue(heap);
  solved(current);
  if(current.blankX > 0){                //UP
    NODE temp;
    temp.blankX = current.blankX-1;
    temp.blankY = current.blankY;
    setNode(&temp, &current, "U");
    setCost(&temp, finalState);
    printf("%d %d\n", temp.cost, temp.level);
    if(temp.cost == temp.level){
      solved(temp);
      return;
    }
    enqueue(heap, temp);
    printf("Went up\n");
  }
  if(current.blankX < (N-1)){                //DOWN
    NODE temp;
    temp.blankX = current.blankX+1;
    temp.blankY = current.blankY;
    setNode(&temp, &current, "D");
    setCost(&temp, finalState);
    if(temp.cost == temp.level){
      solved(temp);
      return;
    }
    enqueue(heap, temp);
    printf("Went down\n");
  }
  if(current.blankY < (N-1)){                //RIGHT
    NODE temp;
    temp.blankX = current.blankX;
    temp.blankY = current.blankY+1;
    setNode(&temp, &current, "R");
    setCost(&temp, finalState);
    if(temp.cost == temp.level){
      solved(temp);
      return;
    }
    enqueue(heap, temp);
    printf("Went right\n");
  }
  if(current.blankY > 0){                //LEFT
    NODE temp;
    temp.blankX = current.blankX;
    temp.blankY = current.blankY-1;
    setNode(&temp, &current, "L");
    setCost(&temp, finalState);
    if(temp.cost == temp.level){
      solved(temp);
      return;
    }
    enqueue(heap, temp);
    printf("Went left\n");
  }
  solveProblem(heap, finalState);
}

void main (){
	int i, j;
  int start[] = {1,2,3,4,5,6,0,8,9,10,7,11,13,14,15,12};
  int end[] = {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,0};
  NODE initial, final;
  HEAP heap;
  heap.size = 0;
  for(i = 0; i < N; i++){
    for(j = 0; j < N; j++){
      initial.state[i][j] = start[i*4+j];
    }
  }
  for(i = 0; i < N; i++){
    for(j = 0; j < N; j++){
      final.state[i][j] = end[i*4+j];
    }
  }
  initial.level = 0;
  initial.blankX = 1;
  initial.blankY = 2;
  initial.pastDirections = "";
  setCost(&initial, final);
  display(initial);
  display(final);
  enqueue(&heap, initial);
  printf("%d\n", heap.size);
  solveProblem(&heap, final);
}


